File: majorityElement.java
//Method 1: Use hashmap key-value, increments value of same key (array elements), check values during loop.
class Solution {
    public int majorityElement(int[] nums) {

        HashMap<Integer, Integer> hashMap = new HashMap<>();
        int majorElement = 0;
        int count = 0;

        for(int i:nums){
            if(!hashMap.containsKey(i)){
                hashMap.put(i,1);
            } else{
                hashMap.put(i, hashMap.get(i)+1);
            }

            if(hashMap.get(i) > count){
                count = hashMap.get(i);
                majorElement = i;
            }
        }
		
        return majorElement;      
    }
}
//Method 2: #Boyer-Moore Algorithm, works in O(N) time complexity and O(1) space complexity.
//Use counter to recognize amount of major element, increase or decrease it traversing array. When it is 0, assign major Element as current element.
class Solution {
    public int majorityElement(int[] nums) {

        int count = 0;
        int majorElement = 0;

        for(int i=0; i<nums.length; i++){
            if(count == 0){
                majorElement = nums[i];
            }
            if(majorElement == nums[i]){
                count++;
            }else{
                count--;
            }
        }
        return majorElement;     
    }
}

File: majorityElement.py
#169. Majority Element
#Given an array nums of size n, return the majority element.
#The majority element is the element that appears more than n/2 times. You may assume that the majority element always exists in the array.

class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        #Boyer-Moore Algorithm, works in O(N) time complexity and O(1) space complexity.
        majorElement = 0
        count = 0

        for i in range(len(nums)):    
            if count == 0:
                majorElement = nums[i]
            if majorElement == nums[i]:
                count +=1
            else:
                count -=1
        print(majorElement)
        return majorElement 

class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        #The next solution is based on dictinary, key-value, time and space complexity is O(n)
        myMap = {} #dictionary, hashmap
        maxValue = 0
        result = 0
        
        for number in nums:
            myMap[number] = 1 + myMap.get(number,0)
            
            if myMap[number] > maxValue:
                result = number
                maxValue = myMap[number]
        
        print(f"result: {result}")
        return result